README.md
# dart-xchg
XCHG dart library


CHANGELOG.md
## 0.1.0-dev

initial release


lib/peer.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:pointycastle/api.dart';
import 'package:pointycastle/asymmetric/api.dart';

import 'network.dart';
import 'network_container.dart';
import 'remote_peer.dart';
import 'request_udp.dart';
import 'rsa.dart';
import 'session.dart';
import 'transaction.dart';
import 'udp_address.dart';
import 'utils.dart';

import 'package:http/http.dart' as http;

class XchgServerProcessor {}

class Peer {
  AsymmetricKeyPair<RSAPublicKey, RSAPrivateKey> keyPair = generateRSAkeyPair();
  bool started = false;
  bool stopping = false;

  DateTime lastNetworkUpdateTime = DateTime.now();
  XchgNetwork? network;

  XchgServerProcessor? processor;

  static int udpStartPort = 42000;
  static int udpEndPort = 42031;

  bool currentUdpPortTrying = false;
  bool currentUdpPortValid = false;

  // Client
  Map<String, RemotePeer> remotePeers = {};

  // Server
  Map<String, Transaction> incomingTransactions = {};
  Map<int, Session> sessionsById = {};
  late Timer _timer;
  bool useLocalRouter = false;

  Peer(AsymmetricKeyPair<RSAPublicKey, RSAPrivateKey>? privKey,
      this.useLocalRouter) {
    print("PEER CREATED!");

    if (privKey == null) {
      AsymmetricKeyPair<RSAPublicKey, RSAPrivateKey> pair =
          generateRSAkeyPair();
      privKey = pair;
    }
    keyPair = privKey;

    requestIncomingFramesFromInternet();
    _timer = Timer.periodic(const Duration(milliseconds: 100), (t) {
      requestIncomingFramesFromInternet();
    });
  }

  void requestIncomingFramesFromInternet() {
    checkNetwork();
    if (network != null) {
      String localAddress = addressForPublicKey(keyPair.publicKey);
      var routers = network!.getNodesAddressesByAddress(localAddress);
      for (var router in routers) {
        requestFramesFromInternet(router);
      }
    }
  }

  void start() {}

  void stop() {
    _timer.cancel();
  }

  int lastReceivedMessageId = 0;
  bool requestingFromInternet = false;

  Future<void> requestFramesFromInternet(String host) async {
    if (requestingFromInternet) {
      return;
    }

    requestingFromInternet = true;
    Uint8List localAddressBS = addressBSForPublicKey(keyPair.publicKey);
    Uint8List getMessagesRequest = Uint8List(16 + 30);
    getMessagesRequest.buffer.asUint64List(0)[0] = lastReceivedMessageId;
    getMessagesRequest.buffer.asUint64List(8)[0] = 1024 * 1024;
    copyBytes(getMessagesRequest, 16, localAddressBS);

    Uint8List res = Uint8List(0);
    try {
      res = await httpCall(host, "r", getMessagesRequest, 20000);
    } catch (err) {
      print("ex: $err");
    }
    if (res.length >= 8) {
      lastReceivedMessageId = res.buffer.asUint64List(0)[0];
      int offset = 8;
      while (offset < res.length) {
        if (offset + 128 < res.length) {
          var tempBuf = res.sublist(offset);
          int frameLen = tempBuf.buffer.asUint32List(0)[0];
          if (offset + frameLen <= res.length) {
            Uint8List frame = res.sublist(offset, offset + frameLen);
            processFrame(null, host, frame);
          } else {
            break;
          }
          offset += frameLen;
        } else {
          break;
        }
      }
    }
    requestingFromInternet = false;
  }

  bool updatingNetwork = false;
  Future<void> updateNetwork() async {
    if (updatingNetwork) {
      return;
    }

    if (useLocalRouter) {
      network = XchgNetwork.localRouter();
      return;
    }
    updatingNetwork = true;
    try {
      network = await networkContainerLoadFromInternet();
      if (network != null) {
        /*print(
            "networkContainerLoadFromInternet ok ${network!.name} source:${network!.debugSource}");*/
      }
    } catch (ex) {
      print("networkContainerLoadFromInternet ex $ex");
    }
    updatingNetwork = false;
  }

  Map<String, http.Client> httpClients = {};

  Future<Uint8List> httpCall(String routerHost, String function,
      Uint8List frame, int timeoutMs) async {
    //print("httpCall $routerHost $function");
    http.Client? client;
    if (httpClients.containsKey(routerHost)) {
      client = httpClients[routerHost];
    } else {
      HttpClient.enableTimelineLogging = false;
      client = http.Client();
      httpClients[routerHost] = client;
    }

    if (client != null) {
      try {
        var req = http.MultipartRequest(
            'POST', Uri.parse("http://$routerHost/api/$function"));
        req.fields['d'] = base64Encode(frame);
        http.Response response = await http.Response.fromStream(
            await client.send(req).timeout(Duration(milliseconds: timeoutMs)));
        if (response.statusCode == 200) {
          var resStr = base64Decode(response.body);
          return resStr;
        }
      } catch (ex) {
        print("exception $function exception: $ex");
      } finally {}
    }
    throw "Exception: status code";
  }

  Future<void> processFrame(
      UdpAddress? sourceAddress, String router, Uint8List frame) async {
    // Min size of frame is 128 bytes
    if (frame.length < 128) {
      return;
    }

    int frameType = frame[8];
    switch (frameType) {
      case 0x10:
        processFrame10(sourceAddress, frame);
        break;
      case 0x11:
        processFrame11(sourceAddress, router, frame);
        break;
      case 0x20:
        processFrame20(sourceAddress, frame);
        break;
      case 0x21:
        processFrame21(sourceAddress, frame);
        break;
      default:
    }
  }

// ----------------------------------------
// Incoming Call - Server Role
// ----------------------------------------

  void processFrame10(UdpAddress? sourceAddress, Uint8List frame) {
    Transaction transaction = Transaction.fromBinary(frame, 0, frame.length);

    List<String> transactionsToRemove = [];
    for (var trKey in incomingTransactions.keys) {
      Transaction? incomingTransaction = incomingTransactions[trKey];
      if (incomingTransaction != null) {
        var diff = DateTime.now().difference(incomingTransaction.dtBegin);
        if (diff.inSeconds > 10) {
          transactionsToRemove.add(trKey);
        }
      }
    }
    for (String trKey in transactionsToRemove) {
      incomingTransactions.remove(trKey);
    }

    Transaction incomingTransaction =
        Transaction(0, "", "", 0, 0, 0, 0, Uint8List(0));

    String incomingTransactionCode =
        sourceAddress.toString() + "-" + transaction.transactionId.toString();
    if (incomingTransactions.containsKey(incomingTransactionCode)) {
      incomingTransaction = incomingTransactions[incomingTransactionCode]!;
    } else {
      incomingTransaction = Transaction(0, "", "", 0, 0, 0, 0, Uint8List(0));
      incomingTransaction.frameType = transaction.frameType;
      incomingTransaction.transactionId = transaction.frameType;
      incomingTransaction.sessionId = transaction.frameType;
      incomingTransaction.offset = 0;
      incomingTransaction.totalSize = transaction.totalSize;
      incomingTransaction.dtBegin = DateTime.now();
      incomingTransaction.srcAddress = transaction.srcAddress;
      incomingTransaction.destAddress = transaction.destAddress;
    }

    if (incomingTransaction.data.length != incomingTransaction.totalSize) {
      incomingTransaction.data = Uint8List(incomingTransaction.totalSize);
    }

    for (int i = 0; i < transaction.data.length; i++) {
      int targetIndex = transaction.offset + i;
      if (targetIndex < incomingTransaction.data.length) {
        incomingTransaction.data[targetIndex] = transaction.data[i];
      }
    }
    incomingTransaction.receivedDataLen += transaction.data.length;

    if (incomingTransaction.receivedDataLen < incomingTransaction.totalSize) {
      return;
    }

    incomingTransactions.remove(incomingTransactionCode);

    {
      // TODO: process Call
    }
  }

  void processFrame11(
      UdpAddress? sourceAddress, String router, Uint8List frame) {
    Transaction transaction = Transaction.fromBinary(frame, 0, frame.length);
    RemotePeer? remotePeer;
    for (RemotePeer peer in remotePeers.values) {
      if (peer.remoteAddress == transaction.srcAddress) {
        remotePeer = peer;
        break;
      }
    }
    if (remotePeer != null) {
      transaction.srcUdpAddr = sourceAddress;
      transaction.srcRouterAddr = router;
      remotePeer.processFrame(transaction);
    }
  }

  // ARP LAN request
  void processFrame20(UdpAddress? sourceAddress, Uint8List frame) {
    String localAddress = addressForPublicKey(keyPair.publicKey);
    Uint8List nonce = frame.sublist(8, 8 + 16);
    Uint8List nonceHash = Uint8List.fromList(sha256.convert(nonce).bytes);
    String requestedAddress = utf8.decode(frame.sublist(8 + 16));
    if (requestedAddress != localAddress) {
      return; // This is not my address
    }
    // Send my public key
    Uint8List publicKeyBS = encodePublicKeyToPKIX(keyPair.publicKey);
    Uint8List signature = rsaSign(keyPair.privateKey, nonceHash);
    Uint8List response = Uint8List(0);
    response.addAll(frame.sublist(0, 8));
    response[0] = 0x21;
    response.addAll(nonce);
    response.addAll(signature);
    response.addAll(publicKeyBS);
    sendFrame(sourceAddress, [response], this);
  }

  // ARP LAN response
  void processFrame21(UdpAddress? sourceAddress, Uint8List frame) {
    try {
      Uint8List receivedPublicKeyBS = frame.sublist(128 + 16 + 256);
      var receivedPublicKey = decodePublicKeyFromPKIX(receivedPublicKeyBS);
      String receivedAddress = addressForPublicKey(receivedPublicKey);
      for (var peer in remotePeers.values) {
        if (peer.remoteAddress == receivedAddress) {
          peer.setRemotePublicKey(
              sourceAddress,
              receivedPublicKey,
              frame.sublist(128, 128 + 16),
              frame.sublist(128 + 16, 128 + 16 + 256));
        }
      }
    } catch (ex) {
      print("Exception (0x21)" + ex.toString());
    }
  }

  void checkNetwork() {
    var diffSec = lastNetworkUpdateTime.difference(DateTime.now()).inSeconds;
    if (network == null || diffSec.abs() > 10) {
      updateNetwork();
      lastNetworkUpdateTime = DateTime.now();
    }
  }

  Future<CallResult> call(String remoteAddress, String authData,
      String function, Uint8List data) async {
    RemotePeer? remotePeer;
    CallResult? res;
    res = CallResult();

    // Update network file
    checkNetwork();

    try {
      // Waiting for socket
      if (remotePeers.containsKey(remoteAddress)) {
        remotePeer = remotePeers[remoteAddress];
      } else {
        remotePeer = RemotePeer(this, remoteAddress, authData, keyPair);
        remotePeers[remoteAddress] = remotePeer;
      }
      res = await remotePeer!.call(function, data);
    } catch (err) {
      return CallResult.createError(err.toString());
    }
    return res;
  }
}


lib/request_udp.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:base32/base32.dart';

import 'peer.dart';
import 'udp_address.dart';

Future<void> sendFrame(
    UdpAddress? address, List<Uint8List> frames, Peer peer) async {
  if (address != null) {
    //requestUDP1(address, frames, peer);
  } else {
    if (peer.network != null) {
      for (var frame in frames) {
        var destAddr = base32
            .encode(Uint8List.fromList(frame.sublist(70, 70 + 30)))
            .toLowerCase();
        for (var addr in peer.network!.getNodesAddressesByAddress(destAddr)) {
          peer.httpCall(addr, "w", frame, 1000).catchError((ex) {
            print("WRITE err = $ex");
          });
          //break;
        }
      }
    }
  }
}

/*Future<void> httpCallWrite(String routerHost, Uint8List frame) async {
  var client = http.Client();
  try {
    var req =
        http.MultipartRequest('POST', Uri.parse("http://$routerHost/api/w"));
    req.fields['d'] = base64Encode(frame);

    http.Response response = await http.Response.fromStream(
        await client.send(req).timeout(const Duration(milliseconds: 1000)));

    if (response.statusCode == 200) {
      //print("WRITE OK");
    }
  } catch (ex) {
    print("WRITE $ex");
  }
  client.close();
}*/

Future<void> requestUDP1(
    UdpAddress address, List<Uint8List> frames, Peer peer) async {
  InternetAddress bindAddr = InternetAddress.anyIPv4;

  if (address.address.rawAddress.length == 16) {
    if (address.address.rawAddress[10] == 255 &&
        address.address.rawAddress[11] == 255) {
      address = UdpAddress(
          InternetAddress.fromRawAddress(
              address.address.rawAddress.sublist(12)),
          address.port);
    }
  }

  var udpSocket = await RawDatagramSocket.bind(bindAddr, 0);
  //udpSocket.broadcastEnabled = true;
  udpSocket.listen(
      (event) {
        if (event == RawSocketEvent.write) {
          for (var fr in frames) {
            for (var i = 0; i < 10; i++) {
              var sentBytes = udpSocket.send(fr, address.address, address.port);
              if (sentBytes == fr.length) {
                break;
              }
              print("Sent Error $sentBytes");
            }
          }
        }
        if (event == RawSocketEvent.read) {
          Datagram? dg = udpSocket.receive();
          if (dg != null) {
            peer.processFrame(UdpAddress(dg.address, dg.port), "UDP", dg.data);
            //result = UdpResponse(UdpAddress(dg.address, dg.port), dg.data);
          }
        }

        if (event == RawSocketEvent.closed) {
          udpSocket.close();
        }
        if (event == RawSocketEvent.readClosed) {
          udpSocket.close();
        }
      },
      cancelOnError: true,
      onDone: () {
        udpSocket.close();
      },
      onError: (err) {
        udpSocket.close();
      });

  await Future.delayed(const Duration(milliseconds: 3000));
  udpSocket.close();
}


lib/transaction.dart
import 'dart:typed_data';
import 'package:base32/base32.dart';

import 'udp_address.dart';
import 'utils.dart';

class Transaction {
  int frameType = 0; // offset: 8
  int transactionId = 0; // offset: 16
  int sessionId = 0; // offset: 24
  int offset = 0; // offset: 32
  int totalSize = 0; // offset: 36
  String srcAddress = ""; // offset: 40
  String destAddress = ""; // offset: 70
  Uint8List data = Uint8List(0); // offset: 128

  bool complete = false;
  String error = "";
  Uint8List response = Uint8List(0);
  DateTime dtBegin = DateTime.now();
  Uint8List result = Uint8List(0);

  // Runtime
  UdpAddress? srcUdpAddr;
  String srcRouterAddr = "";

  int receivedDataLen = 0;

  factory Transaction.fromBinary(Uint8List frame, int offset, int size) {
    frame = frame.sublist(offset);
    Transaction tr = Transaction(0, "", "", 0, 0, 0, 0, Uint8List(0));
    tr.frameType = frame[8];
    tr.transactionId = frame.buffer.asUint64List(16)[0];
    tr.sessionId = frame.buffer.asUint64List(24)[0];
    tr.offset = frame.buffer.asUint32List(32)[0];
    tr.totalSize = frame.buffer.asUint32List(36)[0];

    tr.srcAddress = "#" +
        base32.encode(Uint8List.fromList(frame.sublist(40, 70))).toLowerCase();
    tr.destAddress = "#" +
        base32.encode(Uint8List.fromList(frame.sublist(70, 100))).toLowerCase();

    tr.data = frame.sublist(128, size);
    return tr;
  }

  Transaction(
      int frameType_,
      String srcAddr_,
      String destAddr_,
      int transactrionId_,
      int sessionId_,
      int offset_,
      int totalSize_,
      Uint8List data_) {
    frameType = frameType_;
    srcAddress = srcAddr_;
    destAddress = destAddr_;
    transactionId = transactrionId_;
    sessionId = sessionId_;
    offset = offset_;
    totalSize = totalSize_;
    data = data_;
  }

  List<int> serialize() {
    int frameLen = 128 + data.length;
    Uint8List result = Uint8List(frameLen);
    result.fillRange(0, frameLen, 0);

    result.buffer.asUint32List(0)[0] = frameLen;
    result.buffer.asUint32List(4)[0] = 0; // CRC32
    result[8] = frameType;

    result.buffer.asUint64List(16)[0] = transactionId;
    result.buffer.asUint64List(24)[0] = sessionId;
    result.buffer.asUint32List(32)[0] = offset;
    result.buffer.asUint32List(36)[0] = totalSize;

    var srcAddressBS =
        base32.decode(srcAddress.replaceAll("#", "").toUpperCase());
    var destAddressBS =
        base32.decode(destAddress.replaceAll("#", "").toUpperCase());

    copyBytes(result, 40, srcAddressBS);
    copyBytes(result, 70, destAddressBS);
    copyBytes(result, 128, data);

    return result;
  }

  Uint8List int32bytes(int value) =>
      Uint8List(4)..buffer.asInt32List()[0] = value;
  Uint8List int64bytes(int value) =>
      Uint8List(8)..buffer.asInt64List()[0] = value;
}


lib/remote_peer.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:pointycastle/api.dart';
import 'package:pointycastle/asymmetric/api.dart';

import 'aes.dart';
import 'nonces.dart';
import 'packer.dart';
import 'peer.dart';
import 'request_udp.dart';
import 'rsa.dart';
import 'transaction.dart';
import 'udp_address.dart';
import 'utils.dart';

class RemotePeer {
  Peer peer;
  String remoteAddress;
  String authData;

  AsymmetricKeyPair<RSAPublicKey, RSAPrivateKey> keyPair = generateRSAkeyPair();

  RSAPublicKey? remotePublicKey;

  Nonces nonces = Nonces(100);

  UdpAddress? lanConnectionPoint1;

  bool findingConnection = false;
  bool authProcessing = false;
  Uint8List aesKey = Uint8List(0);
  int sessionId = 0;
  int sessionNonceCounter = 0;
  Map<int, Transaction> outgoingTransactions = {};
  int nextTransactionId = 1;

  RemotePeer(this.peer, this.remoteAddress, this.authData, this.keyPair) {
    nextTransactionId = (DateTime.now().microsecondsSinceEpoch % 1000000);
  }

  static String connectionPointString(UdpAddress? udpAddress) {
    if (udpAddress != null) {
      if (udpAddress.address.rawAddress.length == 16) {
        if (udpAddress.address.rawAddress[10] == 255 &&
            udpAddress.address.rawAddress[11] == 255) {
          udpAddress = UdpAddress(
              InternetAddress.fromRawAddress(
                  udpAddress.address.rawAddress.sublist(12)),
              udpAddress.port);
        }
      }

      return udpAddress.toString();
    }
    return "";
  }

  void setRemotePublicKey(UdpAddress? udpAddress, RSAPublicKey publicKey,
      Uint8List nonce, Uint8List signature) {
    if (!nonces.check(nonce)) {
      return;
    }

    if (!rsaVerify(publicKey, nonce, signature)) {
      return;
    }

    print("setLANConnectionPoint for $remoteAddress is $udpAddress");

    lanConnectionPoint1 = udpAddress;
    remotePublicKey = publicKey;
  }

  String lanConnectionPointString() {
    return connectionPointString(lanConnectionPoint1);
  }

  void processFrame(Transaction transaction) {
    if (transaction.frameType == 0x11) {
      processFrame11(transaction);
    }
  }

  void processFrame11(Transaction transaction) {
    //Transaction transaction = Transaction.fromBinary(frame, 0, frame.length);

    if (outgoingTransactions.containsKey(transaction.transactionId)) {
      var t = outgoingTransactions[transaction.transactionId];
      if (transaction.error == "") {
        if (t!.result.length != transaction.totalSize) {
          t.result = Uint8List(transaction.totalSize);
        }
        for (int i = 0; i < transaction.data.length; i++) {
          int targetOffset = i + transaction.offset;
          if (targetOffset >= 0 && targetOffset < t.result.length) {
            t.result[targetOffset] = transaction.data[i];
          }
        }
        t.receivedDataLen += transaction.data.length;
        if (t.receivedDataLen >= transaction.totalSize) {
          t.complete = true;
        }
      } else {
        t!.result = transaction.data;
        t.error = transaction.error;
        t.complete = true;
      }
    }
  }

  void send0x20(UdpAddress udpAddress) {
    return;

    /*var nonce = nonces.next();
    var addressBS = utf8.encode(remoteAddress);
    var request = Uint8List(8 + 16 + addressBS.length);
    request[0] = 0x20;
    copyBytes(request, 8, nonce);
    copyBytes(request, 8 + 16, Uint8List.fromList(addressBS));
    sendFrame(udpAddress, [request], peer);*/
  }

  Future<void> checkInternetConnectionPoint() async {
    if (remotePublicKey != null) {
      return;
    }

    var nonce = nonces.next();
    var addressBS = utf8.encode(remoteAddress);
    var data = Uint8List(16 + addressBS.length);
    copyBytes(data, 0, nonce);
    copyBytes(data, 16, Uint8List.fromList(addressBS));
    Transaction tr = Transaction(
        0x20, localAddress(), remoteAddress, 0, 0, 0, data.length, data);

    var frame = tr.serialize();
    var frameBS = Uint8List.fromList(frame);
    sendFrame(null, [frameBS], peer);
  }

  Future<void> checkLANConnectionPoint() async {
    var nonce = nonces.next();
    var addressBS = utf8.encode(remoteAddress);
    var data = Uint8List(16 + addressBS.length);
    copyBytes(data, 0, nonce);
    copyBytes(data, 16, Uint8List.fromList(addressBS));

    Transaction tr = Transaction(
        0x20, localAddress(), remoteAddress, 0, 0, 0, data.length, data);

    //socket.broadcastEnabled = true;
    //socket.send(request, InternetAddress("255.255.255.255"), 42000);

    var frame = tr.serialize();

    var frameBS = Uint8List.fromList(frame);

    for (int i = Peer.udpStartPort; i < Peer.udpEndPort; i++) {
      sendFrame(UdpAddress(InternetAddress("127.0.0.1"), i), [frameBS], peer);
    }
  }

  Future<CallResult> call(String function, Uint8List data) async {
    await checkInternetConnectionPoint();

    if (lanConnectionPoint1 == null) {
      if (lanConnectionPoint1 == null) {
        checkLANConnectionPoint();
      }
    }

    // Waiting for network
    for (int i = 0; i < 100; i++) {
      if (peer.network != null) {
        break;
      }
      await Future.delayed(const Duration(milliseconds: 10));
    }
    await checkInternetConnectionPoint();

    // Waiting for public key
    for (int i = 0; i < 100; i++) {
      if (remotePublicKey != null) {
        break;
      }
      await Future.delayed(const Duration(milliseconds: 10));
    }

    if (peer.network == null && lanConnectionPoint1 == null) {
      // No route
      return CallResult.createError("connecting to xchg network");
    }

    if (sessionId == 0) {
      print("auth start");
      String authRes = await auth(lanConnectionPoint1);
      if (authRes.isNotEmpty) {
        CallResult result = CallResult();
        result.error = "auth error:" + authRes;
        sessionId = 0;
        aesKey = Uint8List(0);
        return result;
      }
      print("auth ok");
    }

    CallResult result =
        await regularCall(lanConnectionPoint1, function, data, aesKey);
    return result;
  }

  Future<String> auth(UdpAddress? remoteConnectionPoint) async {
    if (authProcessing) {
      return "auth processing ...";
    }

    try {
      authProcessing = true;
      var getNonceResult = await regularCall(
          remoteConnectionPoint, "/xchg-get-nonce", Uint8List(0), Uint8List(0));
      if (getNonceResult.isError()) {
        authProcessing = false;
        return "get nonce error:" + getNonceResult.error;
      }
      if (getNonceResult.data.length != 16) {
        authProcessing = false;
        return "nonce != 16";
      }

      if (remotePublicKey == null) {
        authProcessing = false;
        return "remotePublicKey == null";
      }

      var authDataBS = utf8.encode(authData);

      var localPublicKeyBS = encodePublicKeyToPKIX(keyPair.publicKey);

      // Prepare auth frame
      Uint8List authFrameSecret = Uint8List(16 + authDataBS.length);
      copyBytes(authFrameSecret, 0, getNonceResult.data);
      copyBytes(authFrameSecret, 16, Uint8List.fromList(authDataBS));
      // Encrypt auth frame
      Uint8List encryptedAuthFrame =
          await rsaEncrypt(remotePublicKey!, authFrameSecret);

      Uint8List authFrame =
          Uint8List(4 + localPublicKeyBS.length + encryptedAuthFrame.length);
      authFrame.buffer.asUint32List(0)[0] =
          localPublicKeyBS.buffer.lengthInBytes;
      copyBytes(authFrame, 4, localPublicKeyBS);
      copyBytes(authFrame, 4 + localPublicKeyBS.length, encryptedAuthFrame);

      CallResult authResult = await regularCall(
          remoteConnectionPoint, "/xchg-auth", authFrame, Uint8List(0));
      if (authResult.isError()) {
        authProcessing = false;
        return "auth error-:" + authResult.error;
      }

      Uint8List authResultDecrypted =
          await rsaDecrypt(keyPair.privateKey, authResult.data);

      if (authResultDecrypted.length != 8 + 32) {
        authProcessing = false;
        return "authResultDecrypted.length != 8 + 32";
      }

      sessionId = authResultDecrypted.buffer.asInt64List(0)[0];
      aesKey = authResultDecrypted.sublist(8);
    } catch (ex) {
      print("******************** auth Exception ee: $ex");
      authProcessing = false;
    }
    authProcessing = false;
    return "";
  }

  Future<CallResult> regularCall(UdpAddress? remoteConnectionPoint,
      String function, Uint8List data, Uint8List aesKey) async {
    CallResult result = CallResult();

    Uint8List functionBS = Uint8List.fromList(utf8.encode(function));

    bool encrypted = false;
    int localSessionNonceCounter = sessionNonceCounter;
    sessionNonceCounter++;

    if (functionBS.length > 255) {
      return CallResult.createError("functionBS.length > 255");
    }

    Uint8List frame = Uint8List(0);

    if (aesKey.length == 32) {
      frame = Uint8List(8 + 1 + functionBS.length + data.length);
      frame.buffer.asInt64List()[0] = localSessionNonceCounter;
      frame[8] = functionBS.length;
      copyBytes(frame, 9, functionBS);
      copyBytes(frame, 9 + functionBS.length, data);
      frame = packBytes(frame);
      frame = aesEncrypt(aesKey, frame);
      encrypted = true;
    } else {
      frame = Uint8List(1 + functionBS.length + data.length);
      frame[0] = functionBS.length;
      copyBytes(frame, 1, functionBS);
      copyBytes(frame, 1 + functionBS.length, data);
    }

    if (!function.contains("data_item_history")) {
      //print("regularCall: $function len:${frame.length} sessionId:$sessionId");
    }

    CallResult callResult =
        await executeTransaction(remoteConnectionPoint, sessionId, frame);
    if (callResult.isError()) {
      return callResult;
    }

    if (encrypted) {
      try {
        callResult.data = aesDecrypt(aesKey, callResult.data);
        callResult.data = unpack(callResult.data);
      } catch (ex) {
        sessionId = 0;
        aesKey = Uint8List(0);
        return CallResult.createError(ex.toString());
      }
    }

    if (callResult.data.isEmpty) {
      return CallResult.createError("callResult.data.length < 1");
    }

    if (callResult.data[0] == 0) {
      // Success
      result = CallResult();
      result.data = callResult.data.sublist(1);

      if (!function.contains("data_item_history")) {
        //print("regularCall ok: $function");
      }

      return result;
    }

    if (callResult.data[0] == 1) {
      // High level Error
      result = CallResult.createError(utf8.decode(callResult.data.sublist(1)));
      return result;
    }

    result = CallResult.createError("Wrong response status");

    return result;
  }

  String localAddress() {
    return addressForPublicKey(keyPair.publicKey);
  }

  Future<CallResult> executeTransaction(
      UdpAddress? remoteConnectionPoint, int sessionId, Uint8List data) async {
    int localTransactionId = nextTransactionId;
    nextTransactionId++;

    Transaction tr = Transaction(0x10, localAddress(), remoteAddress,
        localTransactionId, sessionId, 0, data.length, data);
    outgoingTransactions[tr.transactionId] = tr;

    List<Uint8List> frames = [];

    int offset = 0;
    int blockSize = 524;
    while (offset < tr.data.length) {
      int currentBlockSize = blockSize;
      int restDataLen = tr.data.length - offset;
      if (restDataLen < currentBlockSize) {
        currentBlockSize = restDataLen;
      }

      var blockTransaction = Transaction(
          tr.frameType,
          tr.srcAddress,
          tr.destAddress,
          tr.transactionId,
          tr.sessionId,
          offset,
          tr.totalSize,
          tr.data.sublist(offset, offset + currentBlockSize));

      Uint8List blockFrame = Uint8List.fromList(blockTransaction.serialize());
      frames.add(blockFrame);
      offset += currentBlockSize;
    }
    sendFrame(remoteConnectionPoint, frames, peer);

    for (int i = 0; i < 100; i++) {
      if (tr.complete) {
        outgoingTransactions.remove(localTransactionId);
        if (tr.error.isNotEmpty) {
          return CallResult.createError(tr.error);
        }

        CallResult result = CallResult();
        result.data = tr.result;
        return result;
      }
      await Future.delayed(const Duration(milliseconds: 10));
    }

    outgoingTransactions.remove(localTransactionId);
    resetConnectionPoint();
    return CallResult.createError("transaction timeout");
  }

  void resetConnectionPoint() {
    lanConnectionPoint1 = null;
  }
}
